applications/MappingApplication/custom_mappers/coupling_geometry_mapper.cpp

```diff
@@ -356,14 +356,30 @@ void CouplingGeometryMapper<TSparseSpace, TDenseSpace>::CalculateMappingMatrixWi
         rProjectedInterfaceMatrix.size2());
 
     const size_t n_rows = mpMappingMatrix->size1();
-    Vector solution(n_rows);
-    Vector projector_column(n_rows);
 
-    for (size_t i = 0; i < mpMappingMatrix->size2(); ++i)
-    {
-        for (size_t j = 0; j < n_rows; ++j) projector_column[j] = rProjectedInterfaceMatrix(j, i); // TODO try boost slice or project
-        mpLinearSolver->Solve(rConsistentInterfaceMatrix, solution, projector_column);
-        for (size_t j = 0; j < n_rows; ++j) (*mpMappingMatrix).insert_element(j, i,solution[j]);
+    struct solver_tls {
+        Vector solution;
+        Vector projector_column;
+        LinearSolverSharedPointerType p_lin_solver = nullptr;
+    };
+
+    auto lambda_compute_row = [&](std::size_t i, solver_tls& rTls){
+        // first check if tls is initialized
+        if (!rTls.p_lin_solver) {
+            rTls.p_lin_solver = LinearSolverFactory<TSparseSpace, TDenseSpace>().Create(mMapperSettings["linear_solver_settings"]);
```

Comment:
@peterjwilson I think this needs a bit of a refactor, together with the CreateLinearSolver method

Reply:
what did you have in mind?